src/planner/builder/expr_test.go

```diff
@@ -128,18 +128,74 @@ func TestWhereFilters(t *testing.T) {
 		p, err := scanTableExprs(log, route, database, sel.From)
 		assert.Nil(t, err)
 
-		joins, filters, err := parseWhereOrJoinExprs(sel.Where.Expr, p.getReferTables())
+		p, err = pushFilters(p, sel.Where.Expr)
 		assert.Nil(t, err)
 
-		err = p.pushFilter(filters)
+		_, err = p.calcRoute()
 		assert.Nil(t, err)
 
-		p = p.pushEqualCmpr(joins)
-
-		_, err = p.calcRoute()
 		assert.Nil(t, err)
+	}
+}
+
+func TestWhereFiltersError(t *testing.T) {
+	query := "select * from A where id=0x12"
+	want := "hash.unsupported.key.type:[3]"
+	log := xlog.NewStdLog(xlog.Level(xlog.PANIC))
+	database := "sbtest"
+
+	route, cleanup := router.MockNewRouter(log)
+	defer cleanup()
+
+	err := route.AddForTest(database, router.MockTableMConfig(), router.MockTableBConfig(), router.MockTableGConfig())
+	assert.Nil(t, err)
+
+	node, err := sqlparser.Parse(query)
+	assert.Nil(t, err)
+	sel := node.(*sqlparser.Select)
+
+	p, err := scanTableExprs(log, route, database, sel.From)
+	assert.Nil(t, err)
+
+	// where filter error.
+	{
+		p, err = pushFilters(p, sel.Where.Expr)
+		got := err.Error()
+		assert.Equal(t, want, got)
+	}
+	// check shard error.
+	{
+		_, err = checkShard("B", "id", p.getReferTables(), route)
+		assert.Equal(t, "unsupported: unknown.column.'B.id'.in.field.list", err.Error())
+	}
+	// get on tableinfo.
+	{
+		getOneTableInfo(nil)
+	}
+	// splitAndExpression.
+	{
+		splitAndExpression(nil, nil)
+	}
+}
+
+func TestGetSelectExprs(t *testing.T) {
```

Comment:
It is related to getSelectExprs on the project.go, 
Maybe it is better to modify the position.